File: examples/05-production/go/versioning_behavior.go
package production

import (
	"go.temporal.io/sdk/worker"
	"go.temporal.io/sdk/workflow"
)

// Worker Versioning behavior. Go has no class annotation like Java's
// @WorkflowVersioningBehavior; the behavior is set per workflow type at
// registration via RegisterWorkflowOptions.VersioningBehavior. Pinned keeps
// existing executions on a compatible Build ID; AutoUpgrade lets long-runners
// move to newer compatible worker code.
func RegisterVersionedWorkflows(w worker.Worker) {
	w.RegisterWorkflowWithOptions(ShortLivedCheckoutWorkflow, workflow.RegisterOptions{
		Name:               "ShortLivedCheckoutWorkflow",
		VersioningBehavior: workflow.VersioningBehaviorPinned,
	})
	w.RegisterWorkflowWithOptions(SubscriptionLifecycleWorkflow, workflow.RegisterOptions{
		Name:               "SubscriptionLifecycleWorkflow",
		VersioningBehavior: workflow.VersioningBehaviorAutoUpgrade,
	})
}

// ShortLivedCheckoutWorkflow: existing executions stay on compatible workers.
func ShortLivedCheckoutWorkflow(ctx workflow.Context, cartID string) error {
	return nil
}

// SubscriptionLifecycleWorkflow: long-running executions can auto-upgrade.
func SubscriptionLifecycleWorkflow(ctx workflow.Context, subscriptionID string) error {
	return nil
}

File: examples/05-production/python/versioning_behavior.py
from temporalio import workflow
from temporalio.common import VersioningBehavior

# Worker Versioning behavior. Python sets it per-workflow with the
# `versioning_behavior` argument to @workflow.defn (the analogue of Java's
# @WorkflowVersioningBehavior); the enum lives in temporalio.common. PINNED keeps
# existing executions on a compatible Build ID; AUTO_UPGRADE lets long-runners
# move to newer compatible worker code.


@workflow.defn(versioning_behavior=VersioningBehavior.PINNED)
class ShortLivedCheckoutWorkflow:
    @workflow.run
    async def run(self, cart_id: str) -> None:
        # Existing executions stay on compatible workers during rollout.
        ...


@workflow.defn(versioning_behavior=VersioningBehavior.AUTO_UPGRADE)
class SubscriptionLifecycleWorkflow:
    @workflow.run
    async def run(self, subscription_id: str) -> None:
        # Long-running executions can move to newer compatible worker code.
        ...
